[AIRFLOW-5704] Improve Kind Kubernetes scripts for local testing (#6516)
mistercrunch/Airflow: BREEZE.rst

For supported CI test suites, types of unit tests, and other tests, see `TESTING.rst <TESTING.rst>`_.

Running Tests with Kubernetes in Breeze
=======================================

In order to run Kubernetes in Breeze you can start Breeze with ``--start-kind-cluster`` switch. This will
automatically create a Kind Kubernetes cluster in a ``docker`` (docker-in-docker) container started as part
of Breeze environment. Setting up the Kubernetes cluster takes some time so the cluster continues running
until the environment is stopped with ``--stop-environment`` switch or until ``--recreate-kind-cluster``
switch is used additionally to ``--start-kind-cluster``.

The Kubernetes cluster is automatically initialized and starts Airflow webserver and scheduler

> using latest sources. This is done in two steps. First Airflow images are build in the ``docker`` image,

Reviewer: ```
using the latest sources. This is done in two steps. First Airflow images are built in the ``docker`` image,
```
I'm not sure.

Author: Clarified.

Reviewer: ```diff
-using latest sources. This is done in two steps. First Airflow images are build in the ``docker`` image,
+using the latest sources. This is done in two steps. First Airflow images are built in the ``docker`` image,
```
Small typo. latest => the latest, build=> built

then Airflow applications (``scheduler``/``webserver``) are prepared and deployed at the kind cluster using
the images prepared.

Rebuilding images should be faster every next time we enter the environment but it is attempted every time

> you restart breeze environment. Similarly applications are re-deployed every time the breeze environment

Reviewer: ```diff
-you restart breeze environment. Similarly applications are re-deployed every time the breeze environment
+you restart breeze environment. Similarly, applications are re-deployed every time the breeze environment

```

is restarted.

Building Airflow Images
-----------------------

This is done using ``scripts/ci/in_container/kubernetes/setup_kubernetes.sh`` script:

1. Latest ``apache/airflow:master-pythonX.Y-ci`` images are pulled from DockerHub to the 'docker' container.
2. The image is rebuild to take into account the latest sources mounted to Breeze environment.
3. Scripts required for running the image as Kubernetes cluster are added and

>    ``apache/airflow:master-pythonX.Y-ci-kubernetes``

Reviewer: ```diff
-   ``apache/airflow:master-pythonX.Y-ci-kubernetes``
+   ``apache/airflow:master-pythonX.Y-ci-kubernetes`` image.
```

Author: clarified

For supported CI test suites, types of unit tests, and other tests, see `TESTING.rst <TESTING.rst>`_.

Running Tests with Kubernetes in Breeze
=======================================

To run Kubernetes in Breeze, start Breeze with the ``--start-kind-cluster`` switch. This will
automatically create a Kind Kubernetes cluster in the same ``docker`` engine that the user uses to run Breeze
Setting up the Kubernetes cluster takes some time, so the cluster continues running
until the user stops the cluster explicitly with the  ``--stop-kind-cluster`` switch or until the user runs breeze with the ``--recreate-kind-cluster``
switch (rather than ``--start-kind-cluster``).

The cluster name follows the pattern ``airflow-python-X.Y.Z-vA.B.C`` where X.Y.Z is the Python version,
and A.B.C is kubernetes version. This allows the user to have multiple clusters running at the same
time for different python versions and different kubernetes versions.

> The Control Plane  is available from inside the docker image via <CLUSTER_NAME>-control-plane:6443

Reviewer 2: ```suggestion
The Control Plane is available from inside the docker image via ``<CLUSTER_NAME>-control-plane:6443``
```

host:port, the worker of the kind cluster is available at  <CLUSTER_NAME>-worker
and webserver port for the worker is 30809.

> The Kubernetes Cluster is started but in oder to deploy airflow to Kubernetes cluster you need to:

Reviewer 2: ```suggestion
The Kubernetes Cluster is started but in order to deploy airflow to Kubernetes cluster you need to:
```

1. Build the image.
2. Load it to Kubernetes cluster.
3. Deploy airflow application.

It can be done with single script: ``./scripts/ci/in_container/kubernetes/deploy_airflow_to_kubernetes.sh``

> You can however work separately on image in Kubernetes and the deploying the Airflow app in the cluster.

Reviewer 2: ```suggestion
You can, however, work separately on the image in Kubernetes and deploying the Airflow app in the cluster.
```

Building Airflow Images and Loading them to Kubernetes cluster
--------------------------------------------------------------

This is done using ``./scripts/ci/in_container/kubernetes/docker/rebuild_airflow_image.sh`` script:

1. Latest ``apache/airflow:master-pythonX.Y-ci`` images are rebuilt using latest sources.

> 3. New kubernetes image based on the  ``apache/airflow:master-pythonX.Y-ci`` is build with

Reviewer 2: ```suggestion
2. New Kubernetes image based on the  ``apache/airflow:master-pythonX.Y-ci`` is built with
```

   necessary scripts added to run in kubernetes. The image is tagged with
   ``apache/airflow:master-pythonX.Y-ci-kubernetes`` tag.

> 4. The image is loaded to the kind cluster using ``kind load`` command from kind.

Author: ```suggestion
4. The image is loaded to the kind cluster using ``kind load`` command with ``airflow:latest`` tag.
```

> 4. The image is loaded to the kind cluster using ``kind load`` command

Reviewer 2: ```suggestion
3. The image is loaded to the kind cluster using ``kind load`` command
```

          Python version used for the image. This is always major/minor version.
          One of [ 3.6 3.7 ]. Default is the python3 or python on the path.

  -B, --backend <BACKEND>
          Backend to use for tests - it determines which database is used.
          One of [ sqlite mysql postgres ]. Default: sqlite

  -K, --start-kind-cluster

>           Starts kind kubernetes cluster after entering the environment. The cluster is started using

Reviewer: ```diff
-          Starts kind kubernetes cluster after entering the environment. The cluster is started using
+          Starts kind Kubernetes cluster after entering the environment. The cluster is started using

```

          Kubernetes Mode selected and Kubernetes version specifed via --kubernetes-mode and
          --kubernetes-version flags.

  -Z, --recreate-kind-cluster

>           Recreates kind kubernetes cluster if one has already been created. By default if you do not stop

Reviewer: ```diff
-          Recreates kind kubernetes cluster if one has already been created. By default if you do not stop
+          Recreates kind Kubernetes cluster if one has already been created. By default, if you do not stop

```

>           environment, the kubernetes cluster created for testing is continuously running and when

Reviewer: kubernetes => Kubernetes
